cloudigrade/account/util.py

```diff
@@ -126,80 +135,134 @@ def generate_aws_ami_messages(instances_data, ami_list):
     return messages
 
 
-def _create_exchange_and_queue(queue_name):
+def _get_sqs_queue_url(queue_name):
     """
-    Create a Kombu message Exchange and Queue.
+    Get the SQS queue URL for the given queue name.
+
+    This has the side-effect on ensuring that the queue exists.
 
     Args:
-        queue_name (str): The target queue's name
+        queue_name (str): the name of the target SQS queue
 
     Returns:
-        tuple(kombu.Exchange, kombu.Queue)
+        str: the queue's URL.
 
     """
-    exchange = kombu.Exchange(
-        settings.QUEUE_EXCHANGE_NAME,
-        'direct',
-        durable=True
-    )
-    message_queue = kombu.Queue(
-        queue_name,
-        exchange=exchange,
-        routing_key=queue_name
+    sqs = boto3.client('sqs')
+    try:
+        return sqs.get_queue_url(QueueName=queue_name)['QueueUrl']
+    except ClientError as e:
+        if e.response['Error']['Code'].endswith('.NonExistentQueue'):
+            return sqs.create_queue(QueueName=queue_name)['QueueUrl']
+        raise
+
+
+def _sqs_wrap_message(message):
+    """
+    Wrap the message in a dict for SQS batch sending.
+
+    Args:
+        message (object): message to encode and wrap
+
+    Returns:
+        dict: structured entry for sending to send_message_batch
+
+    """
+    return {
+        'Id': str(uuid.uuid4()),
+        # Yes, the outgoing message uses MessageBody, not Body.
+        'MessageBody': base64.b64encode(
+            jsonpickle.encode(message).encode('utf-8')
+        ).decode('utf-8'),
+    }
+
+
+def _sqs_unwrap_message(sqs_message):
+    """
+    Unwrap the sqs_message to get the original message.
+
+    Args:
+        sqs_message (dict): object to unwrap and decode
+
+    Returns:
+        object: the unwrapped and decoded message object
+
+    """
+    return jsonpickle.decode(
+        base64.b64decode(
```

Comment:
Do we need to encode/decode JSON?  I see a need for binary, but not clear why we are doing this.

Reply:
The AWS SQS message API (at least through boto3) cannot take most raw Python objects. It only works well with strings, and something on the producer and consumer sides need to do the conversion.

I chose `jsonpickle` to do that job because is has slightly better support for common non-primitives like `datetime` objects than the `json` module in the standard Python lib (which only supports a _very_ limited number of primitive types).

```
>>> things = [None, '🎉', 1, 1., 1j, b'123', True, (1, 2), [1, 2], {1, 2, 3}, {'hey': 'arnold!'}, datetime.datetime.now()]
>>>
>>> for thing in things:
...     print(jsonpickle.decode(jsonpickle.encode(thing)))
...
None
🎉
1
1.0
1j
b'123'
True
(1, 2)
[1, 2]
{1, 2, 3}
{'hey': 'arnold!'}
2018-06-14 16:29:51.742549
>>>
>>> for thing in things:
...     try:
...         print(json.loads(json.dumps(thing)))
...     except TypeError:
...         print(f'** json could not dumps {thing}')
...
None
🎉
1
1.0
** json could not dumps 1j
** json could not dumps b'123'
True
[1, 2]
[1, 2]
** json could not dumps {1, 2, 3}
{'hey': 'arnold!'}
** json could not dumps 2018-06-14 16:29:51.742549
>>>
```

Reply:
sorry, the question was worded poorly.  I meant base64 encode JSON not convert JSON string to python.